modules/graceful/releasereopen/releasereopen.go

```diff
@@ -0,0 +1,61 @@
+// Copyright 2023 The Gitea Authors. All rights reserved.
+// SPDX-License-Identifier: MIT
+
+package releasereopen
+
+import (
+	"errors"
+	"sync"
+)
+
+type ReleaseReopener interface {
+	ReleaseReopen() error
+}
+
+type Manager struct {
+	mu      sync.Mutex
+	counter int64
+
+	releaseReopeners map[int64]ReleaseReopener
+}
+
+func (r *Manager) Register(rr ReleaseReopener) (cancel func()) {
+	r.mu.Lock()
+	defer r.mu.Unlock()
+
+	r.counter++
+	currentCounter := r.counter
+	r.releaseReopeners[r.counter] = rr
+
+	return func() {
+		r.mu.Lock()
+		defer r.mu.Unlock()
+
+		delete(r.releaseReopeners, currentCounter)
+	}
+}
+
+func (r *Manager) ReleaseReopen() error {
+	r.mu.Lock()
+	defer r.mu.Unlock()
+
+	var errs []error
+	for _, rr := range r.releaseReopeners {
+		if err := rr.ReleaseReopen(); err != nil {
+			errs = append(errs, err)
+		}
+	}
+	return errors.Join(errs...)
```

Comment:
I'm getting this build error with go 1.19:
```
modules/graceful/releasereopen/releasereopen.go:48:16: undefined: errors.Join
```

It seems this was added in 1.20, while the Gitea `go.mod` lists 1.19 as the go version.
https://pkg.go.dev/errors#Join

Reply:
We once had a build step on CI that built on minimum supported go version, but it appears that may have gotten lost with the migration to GitHub actions.

Reply:
Gitea supports the last two go versions and Go 1.21 is still around 3 months off. If we change this policy, we have to update our docs etc.

Reply:
Hmm, yes, the problem is that if we still use Golang 1.19, then after Golang 1.21 gets released, Golang 1.19 won't get any updates.

So, I think it's good to make Gitea's release can use active Golang version (it's not a must if most people prefer to keep Gitea 1.20 on Golang 1.19)

Reply:
I also remembered that somewhere mentions that "Gitea supports the last two go versions", but I can not find it now .... where is it? 🤣  Has it already been removed?